### DISCUSSION TITLE
sys._MEIPASS doesn't do anything?

### DISCUSSION THREAD

**Discussion by ughstudios** [pyinstaller/pyinstaller#6984] [Help]
Everything I've read on stack overflow points to say that `sys._MEIPASS ` is where some temporary files are "created." I have some questions:

1. What determines if it's created?
2. Why is there nothing in that directory when I look at it?

I want to spawn subprocesses (not using multiprocessing) that will take the current interpreter using `sys.executable` and then append other values.

Imagine this scenario:
```cmdline = [sys.executable, 'myfile.py', '--args=1,2,3']```

This will never work in a pyinstaller build of an application, because `sys.executable` will point to the .exe, not to an interpreter. Also, `myfile.py` won't exist. 

How can I make this work?

**Comment by rokm** [comment]
`sys._MEIPASS` points to the directory containing the frozen application. In `onedir` mode, it points to the directory containing the executable (and all other collected application data). In `onefile` mode, it points to the temporary directory where the collected application data is unpacked. 

> 1. What determines if it's created?

The `onefile` mode.

> 2. Why is there nothing in that directory when I look at it?

*When* are you looking? After the frozen python code exits, the parent process cleans up the temporary directory...


> Imagine this scenario: `cmdline = [sys.executable, 'myfile.py', '--args=1,2,3']`
> 
> This will never work in a pyinstaller build of an application, because `sys.executable` will point to the .exe, not to an interpreter. Also, `myfile.py` won't exist.
> 
> How can I make this work?

By making your entry-point script handle the case when it is passed arguments (assuming the default run mode is without parameters), read the passed .py file and run it (see https://github.com/orgs/pyinstaller/discussions/5978#discussioncomment-971765 for an example). That's assuming that the frozen python environment contains all necessary modules for running that script...

**Reply by ughstudios** [reply]
That's an interesting way of doing it. So with this `--run-script` function you added, you're simply passing the arguments there? So in theory, you can actually just run any script that is packaged into the .exe and it will work? Is `exec()` good practice though?

**Reply by ughstudios** [reply]
Is there a way to obtain an exit code from running `exec()` it does not seem to be possible from what I've read in the documentation for that function

**Reply by rokm** [reply]
> So in theory, you can actually just run any script that is packaged into the .exe and it will work? Is `exec()` good practice though?

This is actually meant primarily for running external scripts (and works only as long as script uses those modules that have been collected into frozen environment).

If you intend to have `myfile.py` collected into the build itself, then I don't see much point in trying to run it as a script. Instead, turn it into a module and call its "main" function from the entry-point script based on some command-line argument. This will also ensure that the code in `myfile.py` will undergo proper dependency analysis.

**Reply by ughstudios** [reply]
So this entire codebase I'm working in is based around the idea of having functions that spawn processes to do work. There's thousands of potential functions. It's core to the entire codebase. 

I learned just now that `execfile` is also a function. Is there a way to access the bundled ('frozen') python scripts? Like what if part of my code does this:

1. Generate command line arguments and paths
2. sys.exec + paths + args
3. Run

In this scenario, the paths for all of my code currently use the `inspect.getmodule()` and `inspect.get_module_src()` functions to get the code. Currently, those functions seem to be returning stuff like this:

os.path.join(os.getcwd(), 'myfile.pyc') such that `myfile.pyc` is actually:
```os.path.abspath(__file__)``` in my actual code, but it's being converted to the cwd and it cannot find it. 

Is there a way to do this?

**Reply by rokm** [reply]
PyInstaller by default collects byte-compiled modules into executable-embedded PYZ archive. But the loader sets `__file__` to point where they would be if they had been on disk as individual files, in order to enable referencing of data files via modules' `__file__`. So you won't find collected modules as files in `sys._MEIPASS` (unless you build with `--debug noarchive`).

**Reply by bwoodsend** [reply]
> So this entire codebase I'm working in is based around the idea of having functions that spawn processes to do work.

But why are insisting on doing this? If you want concurrency, use multiprocessing. If you just want to reuse code between scripts then but the reusable bits into functions/classes then import them from each other. Otherwise, asides from not working with PyInstaller, you'll end up with something that runs horrendously slowly due to each process's startup time, can't easily communicate any message that isn't conveniently string-ify-able and is generally a misery to debug.

**Reply by ughstudios** [reply]
This codebase is older than I am and it supports multiple billion dollar products. I am not going to change this lol. 

@rokm it seems like if I am trying to use `inspect.getsourcefile()` then I cannot use the solution you posted above (https://github.com/orgs/pyinstaller/discussions/5978#discussioncomment-971765)

Is there a way that I can get this to work when using the `inspect` module?

Ideally:
```
module = inspect.getmodule(object)
src = insepect.getsourcefile(module)


if pyinstaller():
    start process that runs what i want it to do
else:
    run normally

```

In this example, `src` is pointing to the path of a `.pyc` file, and trying to open that file, and execute it will not work, because my intent is to execute a python file that is packaged within the .exe

**Reply by ughstudios** [reply]
Do I have to simply include the actual python source code in my pyinstaller app to do this...? It seems to kill the point of using the pyinstaller tool.

**Reply by rokm** [reply]
> Do I have to simply include the actual python source code in my pyinstaller app to do this...? It seems to kill the point of using the pyinstaller tool.

To be fair, nobody claimed that your code architecture is very much suitable for PyInstaller; in fact, both of us have tried to hint that this is decidedly not the case...

If you want to run source files, you will obviously need to collect source files in your application.

If you are fine with running code objects from byte-compiled modules, you can use pyinstaller's collection mechanism. If you absolutely want to load and run .pyc *files*, you will need to ensure that collected modules are not embedded into PYZ archive (using `noarchive` option; or, with upcoming PyInstaller release, you will be able to collect a specific package as .pyc files instead of PYZ archive - #6945). Or you could also retrieve the code object from module's loader, and write it to a file, and then run from it. But then again, why not just retrieve the code object based on module's name (passed via command-line) and run that directly; no need to care whether the module is in a separate file, in a PYZ archive, or wherever. Even if you want to run list all available modules dynamically, you can do so via for example `pkgutil.iter_modules`.

**Reply by ughstudios** [reply]
> But then again, why not just retrieve the code object based on module's name (passed via command-line) and run that directly; no need to care whether the module is in a separate file, in a PYZ archive, or wherever.

How would one do this? You can load a module that's packaged within the .exe, purely by it's name? So let's say I have a module called `animal` that's in `animal.py`, how would I load this module, and call a function on it?

**Reply by ughstudios** [reply]
> > Do I have to simply include the actual python source code in my pyinstaller app to do this...? It seems to kill the point of using the pyinstaller tool.
> 
> To be fair, nobody claimed that your code architecture is very much suitable for PyInstaller; in fact, both of us have tried to hint that this is decidedly not the case...
> 
> If you want to run source files, you will obviously need to collect source files in your application.
> 
> If you are fine with running code objects from byte-compiled modules, you can use pyinstaller's collection mechanism. If you absolutely want to load and run .pyc _files_, you will need to ensure that collected modules are not embedded into PYZ archive (using `noarchive` option; or, with upcoming PyInstaller release, you will be able to collect a specific package as .pyc files instead of PYZ archive - #6945). Or you could also retrieve the code object from module's loader, and write it to a file, and then run from it. But then again, why not just retrieve the code object based on module's name (passed via command-line) and run that directly; no need to care whether the module is in a separate file, in a PYZ archive, or wherever. Even if you want to run list all available modules dynamically, you can do so via for example `pkgutil.iter_modules`.

Out of curiosity, why does multiprocessing just work out of the box, but not simply spawning your own processes?

**Reply by rokm** [reply]
> > But then again, why not just retrieve the code object based on module's name (passed via command-line) and run that directly; no need to care whether the module is in a separate file, in a PYZ archive, or wherever.
> 
> How would one do this? You can load a module that's packaged within the .exe, purely by it's name? 

That's how python's module importing works in general, doesn't it?

> So let's say I have a module called `animal` that's in `animal.py`, how would I load this module, and call a function on it?

<details><summary>program.py</summary><p>

```python
#!/usr/bin/env python3
# program.py
import sys

if __name__ == '__main__':
    if len(sys.argv) == 1:
        # Main program codepath
        import subprocess

        print("Enter name of the module")
        modname = input()
        if not modname:
            print("No module name specified!")
            sys.exit(1)

        if getattr(sys, 'frozen', False):
            # Frozen application: program modname
            subprocess.check_call([sys.executable, modname])
        else:
            # Original python: python program.py modname
            subprocess.check_call([sys.executable, __file__, modname])
    else:
        # Subprocess codepath
        modname = sys.argv[1]
        print(f"Loading module {modname}...")
    
        # Import module
        import importlib
        mod = importlib.import_module(modname)

        # Call function from module
        mod.hello()
```

</p></details>

<details><summary>animal.py</summary><p>

```python
def hello():
   print("Hello from animal!")
```

</p></details>

```
pyinstaller --clean --noconfirm --onefile program.py  --hiddenimport animal
```
(`animal` needs to be added as a hidden import, because it is not directly imported anywhere in the `program.py`).

**Reply by rokm** [reply]
> Out of curiosity, why does multiprocessing just work out of the box, but not simply spawning your own processes?

Because our [`multiprocessing` runtime hook](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/hooks/rthooks/pyi_rth_multiprocessing.py)  monkey-patches `multiprocessing.freeze_support` with  a function that inspects `sys.argv` and diverts the program flow if the arguments correspond to those used by `multiprocessing` helper processes. It is also why multiprocessing does not work if you forget to call `freeze_support` at the start of your program, or why other multiprocessing frameworks are not supported.

**Reply by rokm** [reply]
And spawning your own processes does work, but you need to account for the fact that `sys.executable` is not pointing to `python.exe`, because the runtime environment may not even have python installed...

**Reply by ughstudios** [reply]
> > Out of curiosity, why does multiprocessing just work out of the box, but not simply spawning your own processes?
> 
> Because our [`multiprocessing` runtime hook](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/hooks/rthooks/pyi_rth_multiprocessing.py) monkey-patches `multiprocessing.freeze_support` with a function that inspects `sys.argv` and diverts the program flow if the arguments correspond to those used by `multiprocessing` helper processes. It is also why multiprocessing does not work if you forget to call `freeze_support` at the start of your program, or why other multiprocessing frameworks are not supported.

What if you don't have a function name? What if I want to pass the function name by a string? What if I pass the script, and a function name. I wouldn't be able to just do `mod.hello()` I would want to be able to invoke a function by it's string name.

**Reply by rokm** [reply]
> What if you don't have a function name? What if I want to pass the function name by a string? What if I pass the script, and a function name. I wouldn't be able to just do `mod.hello()` I would want to be able to invoke a function by it's string name.

Perhaps you should go and read about python objects and their attributes. These questions have nothing to do with PyInstaller, and we aren't running a coding bootcamp here...

**Reply by ughstudios** [reply]
> > What if you don't have a function name? What if I want to pass the function name by a string? What if I pass the script, and a function name. I wouldn't be able to just do `mod.hello()` I would want to be able to invoke a function by it's string name.
> 
> Perhaps you should go and read about python objects and their attributes. These questions have nothing to do with PyInstaller, and we aren't running a coding bootcamp here...

I did not realize that import_module() returned the module. I thought it simply created it in the namespace. I was able to figure out my issue.
